share/src/main/java/org/apache/kafka/server/share/persister/PersisterStateManager.java

```diff
@@ -0,0 +1,965 @@
+ * limitations under the License.
+ */
+
+package org.apache.kafka.server.share.persister;
+
+import org.apache.kafka.clients.ClientResponse;
+import org.apache.kafka.clients.CommonClientConfigs;
+import org.apache.kafka.clients.KafkaClient;
+import org.apache.kafka.clients.RequestCompletionHandler;
+import org.apache.kafka.common.Node;
+import org.apache.kafka.common.Uuid;
+import org.apache.kafka.common.internals.Topic;
+import org.apache.kafka.common.message.FindCoordinatorRequestData;
+import org.apache.kafka.common.message.FindCoordinatorResponseData;
+import org.apache.kafka.common.message.ReadShareGroupStateRequestData;
+import org.apache.kafka.common.message.ReadShareGroupStateResponseData;
+import org.apache.kafka.common.message.WriteShareGroupStateRequestData;
+import org.apache.kafka.common.message.WriteShareGroupStateResponseData;
+import org.apache.kafka.common.protocol.ApiKeys;
+import org.apache.kafka.common.protocol.Errors;
+import org.apache.kafka.common.requests.AbstractRequest;
+import org.apache.kafka.common.requests.AbstractResponse;
+import org.apache.kafka.common.requests.FindCoordinatorRequest;
+import org.apache.kafka.common.requests.FindCoordinatorResponse;
+import org.apache.kafka.common.requests.ReadShareGroupStateRequest;
+import org.apache.kafka.common.requests.ReadShareGroupStateResponse;
+import org.apache.kafka.common.requests.WriteShareGroupStateRequest;
+import org.apache.kafka.common.requests.WriteShareGroupStateResponse;
+import org.apache.kafka.common.utils.ExponentialBackoff;
+import org.apache.kafka.common.utils.Time;
+import org.apache.kafka.common.utils.Utils;
+import org.apache.kafka.server.share.SharePartitionKey;
+import org.apache.kafka.server.util.InterBrokerSendThread;
+import org.apache.kafka.server.util.RequestAndCompletionHandler;
+import org.apache.kafka.server.util.timer.Timer;
+import org.apache.kafka.server.util.timer.TimerTask;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Random;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+/**
+ * This class encapsulates various handler classes corresponding to share
+ * state RPCs. It also holds an {@link InterBrokerSendThread} specialization
+ * which manages the sending the RPC requests over the network.
+ * This class is for the exclusive purpose of being used with {@link DefaultStatePersister}
+ * but can be extended for other {@link Persister} implementations as well.
+ */
+public class PersisterStateManager {
```

Comment:
Ok, making sure I understand this class. PersisterStateManager owns a single SendThread which is used to send requests to the ShareCoordinator. The send thread has a queue of "handlers" which are the things responsible building an RPC and handling the response.

Handlers are added to the queue in the following ways:
* by DefaultSharePersister (the "normal" flow)
* periodically by PersisterTimerTask
* recursively by the handlers when the coordinator is not found

We read from the queue when SendThread wakes up and has its generateRequests called by the network layer. We first check the queue for FIND_COORDINATORS and non-batchable RPCs and add those to the list of requests to send. Next we go through the batchable RPCs and add what we can to the list of outgoing requests.

A few questions:

1) What is the lifetime of the handler objects? Are they long lived, or scoped to a single RPC?
2) It seems like requests may be reordered since we are putting them into a map. Is this okay? I think order should be maintained for a given Coordinator + RPC Type + group, so maybe it is.
3) What happens to the pending requests when the coordinator changes?
4) What is the purpose of the periodic wakeup?

Reply:
1. Handler objects remain in scope until the request is completed due to normal response or error. The error can happen , it we get a non-retryable error from the API or max retry attempts are exceeded while retrying a retryable error.
2. no they should not be reordered - the requests will ultimately end up in linkedlists (the ultimate value part of the nested map). The coord+rpc+group are just lookups. Whats inside them is definitely ordered (via linkedlist)
3. They will be sent to the wrong coord - but then retry logic should kick in.
4. InterbrokerSendThread maintains internal kafkaclient which is periodically invoked. It is not very active though. Hence wakeup calls have been added after specific events to force the client to fetch new requests from send thread, if it can. All other specializations of interbrokersend thread in the code base either use the event based (wakeup) strategy or keep calling poll method periodically which does the same thing. Using wakeup removes the need of sleeping.

Reply:
Thanks for the explanations! 

On number 2, if a different RPC type is scheduled, will previous RPCs for that group be flushed?